### DISCUSSION TITLE
How to type the return of interpret function?

### DISCUSSION THREAD

**Discussion by ferjul17** [statelyai/xstate#3130] [Q&A] [answered]
Hi,

Since version 4.29, I have some difficulties to type the return of the function `interpret()`.
Here is a simplified use case:

```typescript
import {createMachine, interpret, Interpreter, MachineConfig, StateMachine} from "xstate";

enum MyState {}

class MyContext {}

interface MyEvent {
  type: string;
  data: number;
}

interface MyStateSchema {
  states: Record<MyState, {}>;
}

class MyMachine {

  private interpreter: Interpreter<MyContext, MyStateSchema, MyEvent>;

  public constructor() {
    const config: MachineConfig<MyContext, MyStateSchema, MyEvent> = {}
    const machine = createMachine(config);
    this.interpreter = interpret(machine);
  }

  public run(): Promise<MyEvent> {
    return new Promise((resolve) => {
      this.interpreter.onDone(() => {
        resolve(this.interpreter.state.event);
      })
      this.interpreter.start();
    })
  }

}
```

The error message returned by typescript is the following:

```
index.ts:23:5 - error TS2322: Type 'Interpreter<MyContext, any, MyEvent, { value: any; context: MyContext; }, TypegenDisabled & { missingImplementations: { actions: never; delays: never; guards: never; services: never; }; } & AllowAllEvents & { ...; }>' is not assignable to type 'Interpreter<MyContext, MyStateSchema, MyEvent, { value: any; context: MyContext; }, TypegenDisabled>'.
  The types of 'machine.initial' are incompatible between these types.
    Type 'string | number | symbol | undefined' is not assignable to type 'MyState | undefined'.
      Type 'string' is not assignable to type 'MyState'.

23     this.interpreter = interpret(machine);
       ~~~~~~~~~~~~~~~~
```

I attempted to type `machine` like that:
```typescript
const machine: StateMachine<MyContext, MyStateSchema, MyEvent>  = createMachine(config);
```

But in this case, I have another error:
```
index.ts:23:5 - error TS2322: Type 'Interpreter<MyContext, MyStateSchema, MyEvent, { value: any; context: MyContext; }, TypegenDisabled & { missingImplementations: { ...; }; } & AllowAllEvents & { ...; }>' is not assignable to type 'Interpreter<MyContext, MyStateSchema, MyEvent, { value: any; context: MyContext; }, TypegenDisabled>'.
  The types of 'machine.getStateNodes' are incompatible between these types.
    Type '(state: StateValue | State<MyContext, MyEvent, any, { value: any; context: MyContext; }, TypegenDisabled & { missingImplementations: { ...; }; } & AllowAllEvents & { ...; }>) => StateNode<...>[]' is not assignable to type '(state: StateValue | State<MyContext, MyEvent, any, { value: any; context: MyContext; }, TypegenDisabled>) => StateNode<...>[]'.
      Types of parameters 'state' and 'state' are incompatible.
        Type 'StateValue | State<MyContext, MyEvent, any, { value: any; context: MyContext; }, TypegenDisabled>' is not assignable to type 'StateValue | State<MyContext, MyEvent, any, { value: any; context: MyContext; }, TypegenDisabled & { missingImplementations: { ...; }; } & AllowAllEvents & { ...; }>'.
          Type 'State<MyContext, MyEvent, any, { value: any; context: MyContext; }, TypegenDisabled>' is not assignable to type 'StateValue | State<MyContext, MyEvent, any, { value: any; context: MyContext; }, TypegenDisabled & { missingImplementations: { ...; }; } & AllowAllEvents & { ...; }>'.
            Type 'State<MyContext, MyEvent, any, { value: any; context: MyContext; }, TypegenDisabled>' is not assignable to type 'State<MyContext, MyEvent, any, { value: any; context: MyContext; }, TypegenDisabled & { missingImplementations: { actions: never; delays: never; guards: never; services: never; }; } & AllowAllEvents & { ...; }>'.
              Types of property 'machine' are incompatible.
                Type 'StateMachine<MyContext, any, MyEvent, { value: any; context: MyContext; }, BaseActionObject, any, TypegenDisabled> | undefined' is not assignable to type 'StateMachine<MyContext, any, MyEvent, { value: any; context: MyContext; }, BaseActionObject, any, TypegenDisabled & { ...; } & AllowAllEvents & { ...; }> | undefined'.
                  Type 'StateMachine<MyContext, any, MyEvent, { value: any; context: MyContext; }, BaseActionObject, any, TypegenDisabled>' is not assignable to type 'StateMachine<MyContext, any, MyEvent, { value: any; context: MyContext; }, BaseActionObject, any, TypegenDisabled & { ...; } & AllowAllEvents & { ...; }>'.
                    Types of property '__TResolvedTypesMeta' are incompatible.
                      Type 'TypegenDisabled' is not assignable to type 'TypegenDisabled & { missingImplementations: { actions: never; delays: never; guards: never; services: never; }; } & AllowAllEvents & { indexedActions: IndexByType<...>; indexedEvents: Record<...> & { ...; }; invokeSrcNameMap: Record<...>; }'.

23     this.interpreter = interpret(machine);
       ~~~~~~~~~~~~~~~~
```

Do you have any idea on how I can type `interpreter`?

I cannot use `any`, it is banned in our repo.

**Comment by Andarist** [comment]
It really would be the easiest to just use `any`, like [here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwMZQKYEMaoLLqQC2ADtUAaOYrKMNGcgSSKptSvNwONQGEIiAzYAHNyAZRiYceQiQC+cflAgg4AIgAeAZwlZVAbjgB6Q3BgBPMKk0AuOABYAdAGYADA5cAoD6iIBXFdhm4pKIsl5IADbomppwgbxMqOrwCGEelKhQ-HiocWYAogBuPikecKYWqLbaUMRCemVwACaY6LZ+IABGmQ1pGVk5ecFYogSoIOiIjdqSNnAASqhI0E0APIHDZKEAfL3hUTF5HDK5CF7lNMCFIf0sVLaMzLSZ62YJWMnsQTqoo-jj6C+RRK5HQRDMoPBu3OcDAvk6EWASDgyyINV8SBg0AAFABKKblcqo7QoviCIS2Y5cBLk17vJJ0IY-P4AoHFJjbOAAXlCjSJfBJE04JG5KLQkipJGxqPJuIahNMhE0DluzygotVrGxQpOcsaaQu8MRyKgviIeNsAAUlCBgJpUK9gRyCYTaL4oEQ4CQAO5wa3KO2obHYtCaCARYr4rmcs4K8owJUqxLUNUOPgAET4Qbx3JjfLjofDxWxCbtSaerEyDhmWAcqHZMD1cbgslx+fjic1VGrElgePlhNb+q8siAA).

I was pretty sure that using `InterpreterFrom` would work, like [here](https://www.typescriptlang.org/play?ssl=22&ssc=24&pln=22&pc=39#code/JYWwDg9gTgLgBAbwMZQKYEMaoLLqQC2ADtUAaOYrKMNGcgSSKptSvNwONQGEIiAzYAHNyAZRiYceQiQZNU1WgoBiUCCHIAVAJ5hUQ1EQAiwAM7oARgBtUAEwC+cfmpBwARAA9TErG4DccAD0gXAwuqimAFxwACwAdADMAAxxAKwAUOmGAK6u2Nrikoj2mUhW6KamcPm88h7wCCXplAr8eKjV2gCiAG6GDelwoeHR3lDEQn6DcLaY6NFEuRYKU00tUG1IHfmFWKIEqCDoiNPeklFwAEqoSNC2ADw7PmTFAHyrmfzZREgwwHxwFAYLD5DgyVAACluAmE0TBXFqgiEj20tSw9XYBWe+3wh3QmN6-XI6CI2mJpK04QMxjMlhstleAEoTkNaNkoERAWhJPCSFC+EjGR90mUKlVQdIuCdpjRgD0iusWFRooxmEooKp1PcwnoIPwucCpJwSO9MkMwNlrMAkIC+GNsr9oBDmQhpkNod5bTChHDJSREcIUWjUBjOrtUDi8QS+kxyWS4CT4zo9NSTOZrHZXnAALzFN1ez1HY0dXNAnnaXmQ6GCqZDIYwQimOKK9U5ijyRSsCFF8FC6ZNc2WqzWuBQb7O6IABRcZlQKMJTCzrrro9Y7M5JAA7nBp+pZxCIWhTBArH1mdml-n643mx2lQo4nwjHxIc6c5eVyujye+hCG2ZbzVVgHzOLA4lQGMYD7T84HsRkr1CG8W2AqA4jOWBnVrOs4P7TJ7CAA), but surprisingly it doesn't. For some reason the returned type doesn't understand that one of the conditional branches should be chosen within `createMyMachine` and it expands a conditional type to a union of both alternatives.

I think I might have local changes that would help you to use `Interpreter` here, gonna play with this more tomorrow and perhaps I'm going to push those changes out.

Question - do you plan to use typegen with your code?

**Reply by ferjul17** [reply]
> Question - do you plan to use typegen with your code?

Nop I don't plan to use typegen.

I found a workaround for now (`interpret(machine) as unknown as Interpreter<MyContext, MyStateSchema, MyEvent>;`) but I'm obviously not happy with it.
If you have a solution, it would be very helpful and appreciated!

**Reply by Andarist** [reply]
I've made a silly mistake yesterday in the example with `InterpreterFrom` (I've used `typeof createMachine` instead of `typeof createMyMachine`). This actually works:
[TS playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwMZQKYEMaoLLqQC2ADtUAaOYrKMNGc3A41AYQiIDNgBzcgSSKo1UVAGJQIIAL5x24kHABEADwDOMTKgUBuOAHpdcGAE8wqFQC44AFgB0AZgAMNgKwAoV6iIBXediMBldSxESXckABt0FRU4P1YBVCV4BFDXSlQodjxUWKMAUQA3T2TXOEMTVEs1KGIuLVK4ABNMdEtvEAAjDPrU9Mzs3MCNfwJUEHREBrUNCzgAJVQkaEaAHj8hrHIUgD4e93YvIiQYYDY4FAwsPwZCEgAKJY5uSxumeM4uNaN4rCT6AKCqBG+DG6H+hWK2wAlJMyrQvFAiOc0BpXvdHh8oXsPEpILBzpFork0TkEO4yjRgAUNBQEtRaBlLPxBAyoGIJCtjKYIOxkZccEYSbtyXAwF4OuFgEhzmxql5jtA7jCyWUyo81DKnlwXnhbiw2B8vj9EnRBoDgaDwUUBNs4ABeEINNWy+DjRgke181GC3VMB4G7hYp2GQgqGx9IRUT0R2h3N16oNlVIU8WS6VQQ5KywABTkwBUqC+EJtsNV8MRcBIAHc4LmJPnUHc7mgVBBwkUYXbbSrVWUYKHw3TIxkbGwACJsRtK+3d4O9lttop3fv5wcs4Qj6ZYGyoa0wRO9uCSKFzkOrmMbqA2aawJX1XvHhqpSRAA)

> Nop I don't plan to use typegen.

I'm curious - is it too complicated for your taste? Are you somehow limited from using it? Or perhaps your use case just doesn't play well with typegen?

**Reply by ferjul17** [reply]
I may be wrong but typegen was introduced recently, we haven't checked exactly what it is. Also, it's still an experimental feature, it's maybe not the best idea to use it for our production ^^'
I'll try to use `InterpreterFrom` but I cannot use `createMyMachine()` like that because the return type is not defined, and it is mandatory in our repository.

**Reply by Andarist** [reply]
> I'll try to use InterpreterFrom but I cannot use createMyMachine() like that because the return type is not defined, and it is mandatory in our repository.

This might be a problem. We are doing quite whacky things with TS that go beyond "day to day" TS and I wouldn't consider all of our types to "manually typeable". That's partially a reason why we have those `InterpreterFrom`, `EventFrom`, `MachineOptionsFrom`, etc types - so you can actually obtain those types with minimal effort. 

While, of course, you can type this stuff manually... I don't recommend it and I really don't want to make this part of the public API, mainly cause the actual structure might somehow change. We are learning TS wizardry on a daily basis and we might have to adjust some of this stuff if required - those types like `InterpreterFrom` are considered a public API though and we have every intention to keep them working.

If you absolutely must use explicit return types then the only thing that comes to my mind that would work... are [instantiation expressions](https://github.com/microsoft/TypeScript/pull/47607) but that feature isn't released yet.

**Reply by ferjul17** [reply]
Thank you for the explanation!
I hadn't time to test `InterpreterFrom`, but I'll do and share with you which solution I choose.

**Reply by ferjul17** [reply]
I had not the time to find a better solution, so I just cast return type of `interpret()`:
```typescript
    this.service = interpret(machine) as unknown as Interpreter<Context, Schema, ClaimEvent>;
```

**Reply by gaberogan** [reply, marked as answer]
`InterpreterFrom<typeof machine>` worked great for me
